Firestore/core/src/firebase/firestore/util/filesystem_win.cc

```diff
@@ -0,0 +1,166 @@
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+
+#include <windows.h>
+
+#include <cerrno>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+bool Exists(const Path& path, bool* is_directory) {
+  DWORD attrs = ::GetFileAttributesW(path.c_str());
+  if (attrs == INVALID_FILE_ATTRIBUTES) {
+    return false;
+  }
+
+  if (is_directory) {
+    *is_directory = attrs & FILE_ATTRIBUTE_DIRECTORY;
+  }
+  return true;
+}
+
+Path TempDir() {
+  // Returns a null-terminated string with a trailing backslash.
+  wchar_t buffer[MAX_PATH + 1];
+  DWORD count = ::GetTempPathW(MAX_PATH, buffer);
+  HARD_ASSERT(count > 0, "Failed to determine temporary directory %s",
+              ::GetLastError());
+
+  return Path::FromUtf16(buffer, count);
+}
+
+namespace detail {
+
+Status CreateDir(const Path& path) {
+  if (::CreateDirectoryW(path.c_str(), nullptr)) {
+    return Status::OK();
+  }
+
+  DWORD error = ::GetLastError();
+  if (error == ERROR_ALREADY_EXISTS) {
+    DWORD attrs = ::GetFileAttributesW(path.c_str());
+    if (attrs == INVALID_FILE_ATTRIBUTES) {
+      error = ::GetLastError();
+
+    } else if (attrs & FILE_ATTRIBUTE_DIRECTORY) {
+      return Status::OK();
+
+    } else {
+      return Status{
+          FirestoreErrorCode::FailedPrecondition,
+          StringFormat(
+              "Could not create directory %s: non-directory already exists",
+              path.ToUtf8String())};
+    }
+  }
+
+  return Status::FromLastError(
+      error,
+      StringFormat("Could not create directory %s", path.ToUtf8String()));
+}
+
+Status DeleteDir(const Path& path) {
+  if (RemoveDirectoryW(path.c_str())) {
+    return Status::OK();
+  }
+
+  DWORD error = GetLastError();
+  if (error == ERROR_FILE_NOT_FOUND || error == ERROR_PATH_NOT_FOUND) {
+    return Status::OK();
+  }
+
+  return Status::FromLastError(
+      error,
+      StringFormat("Could not delete directory %s", path.ToUtf8String()));
+}
+
+Status DeleteFile(const Path& path) {
+  if (DeleteFileW(path.c_str())) {
+    return Status::OK();
+  }
+
+  DWORD error = GetLastError();
+  if (error == ERROR_FILE_NOT_FOUND || error == ERROR_PATH_NOT_FOUND) {
+    return Status::OK();
+  }
+
+  return Status::FromLastError(
+      error, StringFormat("Could not delete file %s", path.ToUtf8String()));
+}
+
+/**
+ * Recursively deletes a path known to be a directory.
+ */
+Status RecursivelyDeleteDir(const Path& parent) {
+  Status result;
+  auto fail = [&](DWORD error) {
+    result.Update(Status::FromLastError(
+        error,
+        StringFormat("Could not delete directory: %s", parent.ToUtf8String())));
+    return result;
+  };
+
+  WIN32_FIND_DATAW find_data;
+  Path pattern = parent.AppendUtf16(L"*", 1);
```

Comment:
If I pass in "c:\path\to\dir" wouldn't this function also remove "c:\path\to\dir_which_is_a_file_that_shares_a_prefix" ?

Other options to using wildcards might be shfileoperation (pre-vista) or ifileoperation (vista+). But the way you've got it actually seems to be more common?

If you just pass in the directory (without '*'), does it do the right thing? Otherwise, I think you'd have to append "/*" rather than just "*". (You might also have to perform some funny business with cleaning up slashes/backslashes. i.e. if the user specified "c:\path\to\delete\" and you append "\*" you'd end up with double backslashes. No idea if that's ok or not.)